### PR TITLE
Fix bugs causing red indexes with remote indexes during translog upload & store recovery

### LINKED ISSUE #10398: [BUG] Shard failure due to fail engine during trimUnreferencedTranslogFiles in InternalEngine for Remote indexes
**Describe the bug**
There is an async task called `AsyncTrimTranslogTask` in IndexService.java (mentioned below as code snippet). The setting `index.translog.retention.check_interval` governs that the periodicity at which the trimming happens. The default value for this setting is 10 minutes. If there are interim issues with remote store, then `RemoteFsTranslog.prepareAndUpload(..)` will throw exception and cause engine failure. Tracing the original commit, it was added to clean up stale translog/ckp files which are unreferenced. This is a valid concern for non-remote indexes as there is a possibility that the translog files can be really huge. This will not happen due to translog generation rollover happening after every indexing operation keeping the file size in check.
   
https://github.com/opensearch-project/OpenSearch/blob/211578ce8ea0259d23d823652ba549bb7e66350e/server/src/main/java/org/opensearch/index/IndexService.java#L1177-L1207

https://github.com/opensearch-project/OpenSearch/blob/211578ce8ea0259d23d823652ba549bb7e66350e/server/src/main/java/org/opensearch/index/translog/InternalTranslogManager.java#L215C1-L231C1

**To Reproduce**
If the remote store interactions during trimming fails at the time of the above mentioned async task run, it will lead to shard failure.

**Expected behavior**
Shard should not fail. 

**Plugins**
Please list all plugins currently enabled.

**Screenshots**

**Host/Environment (please complete the following information):**

**Additional context**

### PR DESCRIPTION
**Pull request by ashking94**
### Description
This PR fixes 2 issues which are causing red indexes while I performed stress testing -
1. https://github.com/opensearch-project/OpenSearch/issues/10400
2. https://github.com/opensearch-project/OpenSearch/issues/10398

### Related Issues
Resolves #10398, #10400.

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ashking94** [comment]
These issues have been found while doing stress testing and It's a bit tricky to simulate these kind of failures in integ tests. Still exploring if there is any way to write an IT for the above fixes.

**Review by gbbafna — APPROVED** [review]

**Comment by ashking94** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/27280/
> * **CommitID:** [8674a1d](https://github.com/opensearch-project/OpenSearch/commit/8674a1d2727425a902f4810ec378f808c7504925)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky tests - https://github.com/opensearch-project/OpenSearch/issues/9859, https://github.com/opensearch-project/OpenSearch/issues/10459, https://github.com/opensearch-project/OpenSearch/issues/10046, https://github.com/opensearch-project/OpenSearch/issues/10154

**Comment by sachinpkale** [comment]
Please add ITs around the changes.

**Comment by ashking94** [comment]
> Please add ITs around the changes.

@sachinpkale  Done. Pls take a relook. Ty!

**Comment by ashking94** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/27395/
> * **CommitID:** [4657a37](https://github.com/opensearch-project/OpenSearch/commit/4657a37975b0ceebcb895daf3445b6d20dc2df14)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Failures -
```
org.opensearch.script.expression.MoreExpressionIT.testSpecialValueVariable {p0={"search.concurrent_segment_search.enabled":"true"}}
org.opensearch.index.IndexServiceTests.testAsyncTranslogTrimTaskOnClosedIndex
```
Of the above, https://github.com/opensearch-project/OpenSearch/issues/10079  is already a reported flaky test.
On the failure `testAsyncTranslogTrimTaskOnClosedIndex`, I have made no changes that should affect the specific test. The change made here is only applicable for remote indexes and does not affect the `IndexServiceTests.testAsyncTranslogTrimTaskOnClosedIndex` test. I am running the specific test on main to see if this fails after high iteration count or low. I have already this test now for more than 200+ iterations with this change and I dont see the failure yet.

**Comment by ashking94** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/27399/
> * **CommitID:** [e0c59a6](https://github.com/opensearch-project/OpenSearch/commit/e0c59a635ee927c697547cdeffde50e87b068dd7)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky tests

File: server/src/main/java/org/opensearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -2314,7 +2317,7 @@ public void openEngineAndRecoverFromTranslog() throws IOException {
         };
 
         // Do not load the global checkpoint if this is a remote snapshot index
-        if (indexSettings.isRemoteSnapshot() == false) {
+        if (indexSettings.isRemoteSnapshot() == false && indexSettings.isRemoteTranslogStoreEnabled() == false) {
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/index/shard/IndexShard.java:2320]: nit: We can use the same check that we used in `trimTranslog` - `isRemoteTranslogEnabled()`

**Review by sachinpkale — APPROVED** [review]

**Pull request merged**